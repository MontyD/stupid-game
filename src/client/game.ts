import { GameRunState } from '../models/entities/game';
import { ObjectOfAny } from '../utils/types';

export class Game {

    public static from(gameEntity: ObjectOfAny): Game {
        const game: ObjectOfAny = new this();
        const gameKeys = Object.keys(game);
        Object.keys(gameEntity).forEach(key => {
            if (gameKeys.includes(key)) {
                game[key]  = gameEntity[key];
            }
        });
        return game as Game;
    }

    public runState: GameRunState  = GameRunState.WAITING_FOR_PLAYERS_TO_JOIN;
    public code: string = '';
}
