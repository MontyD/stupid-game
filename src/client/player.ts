import { ObjectOfAny } from '../utils/types';
import { TypeOfPlayer } from '../models/entities/player';

export class Player {

    public static from(playerEntity: ObjectOfAny): Player {
        const player: ObjectOfAny = new Player();
        const playerKeys = Object.keys(player);
        Object.keys(playerEntity).forEach(key => {
            if (playerKeys.includes(key)) {
                player[key]  = playerEntity[key];
            }
        });
        return player as Player;
    }

    public id: string = '';
    public name: string = '';
    public type: TypeOfPlayer = TypeOfPlayer.OBSERVER;
    public isHost: boolean = false;
}
