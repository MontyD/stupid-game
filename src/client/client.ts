import { TopLevelClientToServerMessages, TopLevelServerToSingleClientMessages } from "../models/messages/top-level";
import { ObjectOfAny } from '../utils/types';
import { BroadcastGameMessages, SingleClientGameMessages, ClientToServerGameMessages } from '../models/messages/game';
import { JoinGameOptions } from "../controllers/game-controller";
import { PlayerDTO } from "../models/entities/player";
import { GameDTO, GameRunState } from "../models/entities/game";
import { GameDefinitionDTO } from "../models/entities/game-definition";
import { ValidationError } from "../controllers/validation-error";

export type RequestCommands = TopLevelServerToSingleClientMessages | BroadcastGameMessages | SingleClientGameMessages;

type PlayerHandler = (player: PlayerDTO) => void;
type GameDefinitionHandler = (gameDefinition: GameDefinitionDTO) => void;

export class Client {

    public static createAsHost(socket: SocketIOClient.Socket): Promise<Client> {
        return (new Client(socket)).requestGameAsHost();
    }

    public static joinGameAsPlayer(socket: SocketIOClient.Socket, options: JoinGameOptions): Promise<Client> {
        return (new Client(socket)).joinGameAsPlayer(options);
    }

    public game?: GameDTO;
    public gameDefinition?: GameDefinitionDTO;
    public player?: PlayerDTO;
    public otherPlayers: PlayerDTO[] = [];

    private playerJoinedHandlers: PlayerHandler[] = [];
    private playerLeftHandlers: PlayerHandler[] = [];
    private startGameHandlers: GameDefinitionHandler[] = [];

    private constructor(private socket: SocketIOClient.Socket) { }

    public async startGame() {
        this.socket.emit(ClientToServerGameMessages.START);
        await this.takeNext(BroadcastGameMessages.STARTED);
    }

    public onPlayerJoined(handler: PlayerHandler) {
        this.playerJoinedHandlers.push(handler);
    }

    public onPlayerLeft(handler: PlayerHandler) {
        this.playerLeftHandlers.push(handler);
    }

    public onGameStarted(handler: GameDefinitionHandler) {
        this.startGameHandlers.push(handler);
    }

    public dispose(): void {
        this.socket.disconnect();
    }

    private async requestGameAsHost(): Promise<this> {
        this.socket.emit(TopLevelClientToServerMessages.CREATE_GAME);

        const { game, player, gameDefinition } = await this.takeNext(BroadcastGameMessages.CREATED);
        this.game = game;
        this.player = player;
        this.gameDefinition = gameDefinition;
        this.setupHandlers();
        return this;
    }

    private async joinGameAsPlayer(options: JoinGameOptions): Promise<this> {
        this.socket.emit(TopLevelClientToServerMessages.JOIN_GAME, options);

        const {
             game, player, gameDefinition, otherPlayers,
        } = await this.takeNext(SingleClientGameMessages.JOIN_SUCCESSFUL);
        this.game = game;
        this.player = player;
        this.gameDefinition = gameDefinition;
        this.otherPlayers = otherPlayers;
        this.setupHandlers();
        return this;
    }

    private handlePlayerJoined({ game, player }: { game: GameDTO, player: PlayerDTO }): void {
        this.game = game;
        this.otherPlayers.push(player);
        this.playerJoinedHandlers.forEach(handler => handler(player));
    }

    private handlePlayerLeft({ player: gonePlayer }: { player: PlayerDTO}): void {
        this.otherPlayers = this.otherPlayers.filter(existingPlayer => existingPlayer.id !== gonePlayer.id);
        this.playerLeftHandlers.forEach(handler => handler(gonePlayer));
    }

    private handleGameStarted({ gameDefinition }: { gameDefinition: GameDefinitionDTO }): void {
        if (!this.game) {
            return;
        }
        this.game.runState = GameRunState.RUNNING_ROUND;
        this.gameDefinition = gameDefinition;
        this.startGameHandlers.forEach(handler => handler(gameDefinition));
    }

    private async takeNext(command: RequestCommands): Promise<ObjectOfAny> {
        const result = await Promise.race([
            this.handlerToKeyedPromise<ObjectOfAny>('success', command),
            this.handlerToKeyedPromise<string>(
                'validationError',
                TopLevelServerToSingleClientMessages.VALIDATION_ERROR
            ),
            this.handlerToKeyedPromise<string>('failure', TopLevelServerToSingleClientMessages.ERROR),
        ]);

        if (result.validationError) {
            throw new ValidationError(result.validationError as string);
        }
        if (result.failure) {
            throw new Error(result.failure as string);
        }
        return result.success as ObjectOfAny;
    }

    private handlerToKeyedPromise<T>(keyName: string, eventName: string): Promise<{ [key: string]: T }> {
        return new Promise((resolve) => this.socket.once(eventName, (arg: any) => resolve({ [keyName]: arg })));
    }

    private setupHandlers() {
        this.socket.on(BroadcastGameMessages.PLAYER_JOINED, this.handlePlayerJoined.bind(this));
        this.socket.on(BroadcastGameMessages.PLAYER_DISCONNECTED, this.handlePlayerLeft.bind(this));
        this.socket.on(BroadcastGameMessages.STARTED, this.handleGameStarted.bind(this));
    }

}
