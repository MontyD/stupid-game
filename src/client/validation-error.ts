import { RequestCommands } from './client';

export class ValidationError extends Error {

    public command: RequestCommands;

    constructor(message: string, command: RequestCommands) {
        super(message);
        this.command = command;
    }

}
