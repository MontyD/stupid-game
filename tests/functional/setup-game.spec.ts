import { createClientSocket } from "./util/create-connection";
import { Client } from '../../src/client/client';
import 'jest';

describe('game setup', () => {

    let toDisconnect: SocketIOClient.Socket[] = [];

    const createSocket = (): SocketIOClient.Socket => {
        const newSocket = createClientSocket();
        toDisconnect.push(newSocket);
        return newSocket;
    };

    afterEach(() => {
        toDisconnect.forEach(io => io.disconnect());
        toDisconnect = [];
    });

    it('will create a game', async () => {
        const {game, player, otherPlayers} = await Client.createAsHost(createSocket());
        expect(game!.code.length).toEqual(5);
        expect(game!.runState).toEqual('WAITING_FOR_PLAYERS_TO_JOIN');
        expect(player).not.toBeUndefined();
        expect(player!.name).toEqual('HOST');
        expect(player!.isHost).toBe(true);
        expect(otherPlayers).toHaveLength(0);
    });

    it('will allow players to join the game', async () => {
        const playerJoinedListener = jest.fn();

        const host = await Client.createAsHost(createSocket());
        host.onPlayerJoined(playerJoinedListener);
        expect(host.otherPlayers).toEqual([]);

        const activeClient = await Client.joinGameAsPlayer(createSocket(), {
            playerName: 'new-player',
            gameCode: host.game!.code,
        });

        if (!activeClient.game || !host.game || !activeClient.player || !host.player) {
            throw new Error('game or player was null on activeClient or host');
        }

        expect(activeClient.game.id).toEqual(host.game.id);
        expect(activeClient.otherPlayers).toHaveLength(1);
        expect(activeClient.otherPlayers[0].id).toEqual(host.player.id);
        expect(activeClient.player.name).toEqual('new-player');

        expect(host.otherPlayers).toHaveLength(1);
        expect(host.otherPlayers[0]!.id).toEqual(activeClient.player.id);
        expect(host.otherPlayers[0]!.name).toEqual('new-player');

        expect(playerJoinedListener).toHaveBeenCalledWith(host.otherPlayers[0]);
    });

    it('will handle players disconnecting', async (done) => {
        const socketToDisconnect = createClientSocket();
        const host = await Client.createAsHost(createSocket());
        const activePlayerClient = await Client.joinGameAsPlayer(socketToDisconnect, {
            playerName: 'i-will-leave',
            gameCode: host.game!.code,
        });

        host.onPlayerLeft((player: any) => {
            expect(player.id).toEqual(activePlayerClient.player!.id);
            expect(host.otherPlayers).toEqual([]);
            done();
        });

        socketToDisconnect.disconnect();
    });

    it('will limit the number of players to 12', async () => {
        const host = await Client.createAsHost(createSocket());
        const players = [];

        let i = 0;
        while (i++ < 12) {
            const newPlayer = await Client.joinGameAsPlayer(createSocket(), {
                playerName: `player ${i}`,
                gameCode: host.game!.code,
            });
            players.push(newPlayer);
        }

        await expect((async () => {
            await Client.joinGameAsPlayer(createSocket(), {
                playerName: 'one player too many',
                gameCode: host.game!.code,
            });
        })()).rejects.toMatchSnapshot();
    });

    it('will only allow the game to start if we have enough players', async () => {
        const host = await Client.createAsHost(createSocket());

        let i = 0;
        while (i++ < 2) {
            await expect((async () => {
                const client = await Client.joinGameAsPlayer(createSocket(), {
                    playerName: `not enough players ${i}`,
                    gameCode: host.game!.code,
                });
                await client.startGame();
            })()).rejects.toMatchSnapshot();
        }
    });

    it('will not start the same game twice', async () => {
        const host = await Client.createAsHost(createSocket());

        let i = 0;
        while (i++ < 3) {
            await Client.joinGameAsPlayer(createSocket(), {
                playerName: `player ${i}`,
                gameCode: host.game!.code,
            });
        }

        await host.startGame();

        expect((async () => {
            await host.startGame();
        })()).rejects.toThrowErrorMatchingSnapshot();
    });

});
