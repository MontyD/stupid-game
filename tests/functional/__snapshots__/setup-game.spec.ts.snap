// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`game setup will limit the number of players to 12 1`] = `[ValidationError: Game is full]`;

exports[`game setup will not start the same game twice 1`] = `[ValidationError: Not enough players to start game]`;

exports[`game setup will not start the same game twice 2`] = `[ValidationError: Not enough players to start game]`;

exports[`game setup will only allow the game to start if we have enough players 1`] = `[ValidationError: Not enough players to start game]`;

exports[`game setup will only allow the game to start if we have enough players 2`] = `[ValidationError: Not enough players to start game]`;
